[Skia] Fix support for custom cursors with GTK

TingPing opened the pull request:
#### ee489d263ec64f9183dba00baba1b9feb9ec01a5
<pre>
[Skia] Fix support for custom cursors with GTK
<a href="https://bugs.webkit.org/show_bug.cgi?id=271613">https://bugs.webkit.org/show_bug.cgi?id=271613</a>

Reviewed by Adrian Perez de Castro.

* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
(WebCore::skiaImageToGdkPixbuf):
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):

destroy notify was not being called.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::~GraphicsContextSkia):

Canonical link: <a href="https://commits.webkit.org/276737@main">https://commits.webkit.org/276737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ab2574d06e4b832e7bd7fb18c6cec6976b669d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47920 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48055 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28741 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21908 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37227 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39165 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18331 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40243 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41684 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49775 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20375 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16900 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44272 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21683 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43090 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10123 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22042 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21370 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -73,6 +73,17 @@ RefPtr<cairo_surface_t> skiaImageToCairoSurface(SkImage& image)
+    if (!image.peekPixels(&pixmap))
+        return { };
+
+    return adoptGRef(gdk_pixbuf_new_from_data(pixmap.addr8(), GDK_COLORSPACE_RGB, !pixmap.computeIsOpaque(), pixmap.info().bytesPerPixel() * 8, pixmap.width(), pixmap.height(), pixmap.rowBytes(), [](guchar*, void *data) {
```
According to the [documentation](https://docs.gtk.org/gdk-pixbuf/ctor.Pixbuf.new_from_data.html), `has_alpha` signifies “Whether the data has an opacity channel” but Skia's `::computeIsOpaque()` returns “true if all pixels are opaque”. I think this should use `SkPixmap::isOpaque()` which checks whether the image has alpha or not, regardless of whether any pixel is itself opaque, and this seems to match the expectation of the `gdk_pixbuf_new_from_data()` function better.

```suggestion
    return adoptGRef(gdk_pixbuf_new_from_data(pixmap.addr8(), GDK_COLORSPACE_RGB, !pixmap.isOpaque(), pixmap.info().bytesPerPixel() * 8, pixmap.width(), pixmap.height(), pixmap.rowBytes(), [](guchar*, void *data) {
```

Also, a couple of notes more:

- Using `SkPixmap::data8()` asserts that `::bytesPerPixel() == 1` so it will fail for anything that is not 8 bits per component if Skia is ever built with assertions enabled. We should use `::addr(0, 0)` if we want to support images with more than 8bpc.
- Alternatively, we may want to support only a fixed set of formats, and convert to `RGB{A,X}_8888` otherwise.

TingPing replied:
For now ill just assert 8bit since GTK4 only supports that atm.

aperezdc replied:
> For now ill just assert 8bit since GTK4 only supports that atm.

Sounds fine by me. Maybe add a FIXME/TODO stating that we might want to consider converting the incoming image to to 8bpc (`RGB{A,X}_8888`) if needed.

aperezdc reviewed (changes requested):
I left one comment, please take a look.

aperezdc commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -73,6 +73,18 @@ RefPtr<cairo_surface_t> skiaImageToCairoSurface(SkImage& image)
+    if (!image.peekPixels(&pixmap))
+        return { };
+
+    // FIXME: This only supports RGB{A,X}_8888, so we could convert the SkImage if it was in a different format.
```
I would add a check here as well, and return an empty image. Better safe than crashy!

```suggestion
    // FIXME: This only supports RGB{A,X}_8888, so we could convert the SkImage if it was in a different format.
    if (image.colorType() != SkColorType::kRGBA_8888_SkColorType && image.colorType() != SkColorType::kRGB_888x_SkColorType)
        return { };
```

aperezdc commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -73,6 +73,18 @@ RefPtr<cairo_surface_t> skiaImageToCairoSurface(SkImage& image)
+        return { };
+
+    // FIXME: This only supports RGB{A,X}_8888, so we could convert the SkImage if it was in a different format.
+    return adoptGRef(gdk_pixbuf_new_from_data(pixmap.addr8(), GDK_COLORSPACE_RGB, !pixmap.isOpaque(), 8, pixmap.width(), pixmap.height(), pixmap.rowBytes(), [](guchar*, void *data) {
```
...and then here you want `::addr32()` because that returns a 32-bit per pixel pointer (4 bytes per pixel, 8bits per color component). Sorry if I wasn't clear with this in my earlier comment.

```suggestion
    return adoptGRef(gdk_pixbuf_new_from_data(pixmap.addr32(), GDK_COLORSPACE_RGB, !pixmap.isOpaque(), 8, pixmap.width(), pixmap.height(), pixmap.rowBytes(), [](guchar*, void *data) {
```

TingPing replied:
I'm not sure I understand, `addr8()` seems correct here.

aperezdc replied:
The `addr8()` accessor is used for 8bpp (bits per pixel) images (grayscale or indexed color). The documentation in `SkPixmap.h` is clear about this:

```cpp
    /** Returns readable base pixel address. Result is addressable as unsigned 8-bit bytes.
        Will trigger an assert() if SkColorType is not kAlpha_8_SkColorType or
        kGray_8_SkColorType, and is built with SK_DEBUG defined.

        One byte corresponds to one pixel.

        @return  readable unsigned 8-bit pointer to pixels
    */
```

What we want here is the 32bpp (4 bytes per pixel, 1 byte for each of `RGB{A,X}`) accessor, which is `addr32()`:

```cpp
    /** Returns readable base pixel address. Result is addressable as unsigned 32-bit words.
        Will trigger an assert() if SkColorType is not kRGBA_8888_SkColorType or
        kBGRA_8888_SkColorType, and is built with SK_DEBUG defined.

        One word corresponds to one pixel.

        @return  readable unsigned 32-bit pointer to pixels
    */
```

I hope it's clearer now.

TingPing replied:
So the GdkPixbuf constructor doesn't support any format that isn't 8 bits.

However you can construct a GdkTexture/cairo_surface from it and then create a GdkPixbuf off that, so I took this approach.

aperezdc reviewed (commented):
@TingPing my apologies for the churn, I think in my earlier review I wasn't clear about needing to change the accessor used for the pixel data 😇

aperezdc reviewed (approved):
> However you can construct a GdkTexture/cairo_surface from it and then create a GdkPixbuf off that, so I took this approach.

@TingPing That's a clever solution, and I like it that makes the code somewhat more consistent and easier to follow. Thanks!

TingPing committed:
[Skia] Fix support for custom cursors with GTK
https://bugs.webkit.org/show_bug.cgi?id=271613

Reviewed by Adrian Perez de Castro.

* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
(WebCore::skiaImageToGdkPixbuf):
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):

destroy notify was not being called.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::~GraphicsContextSkia):

Canonical link: https://commits.webkit.org/276737@main
